src/lib/rocksdb/database.ml

```diff
@@ -75,23 +79,96 @@ let to_alist t : (Bigstring.t * Bigstring.t) list =
   in
   loop []
 
+let to_bigstring = Bigstring.of_string
+
 let%test_unit "to_alist (of_alist l) = l" =
-  Quickcheck.test
+  Async.Thread_safe.block_on_async_exn
+  @@ fun () ->
+  Async.Quickcheck.async_test ~trials:50
     Quickcheck.Generator.(
       tuple2 String.quickcheck_generator String.quickcheck_generator |> list)
     ~f:(fun kvs ->
-      File_system.with_temp_dir "/tmp/coda-test" ~f:(fun directory ->
-          let s = Bigstring.of_string in
-          let sorted =
-            List.sort kvs ~compare:[%compare: string * string]
-            |> List.map ~f:(fun (k, v) -> (s k, s v))
-          in
-          let db = create directory in
-          List.iter sorted ~f:(fun (key, data) -> set db ~key ~data) ;
-          let alist =
-            List.sort (to_alist db)
-              ~compare:[%compare: Bigstring.t * Bigstring.t]
+      match Hashtbl.of_alist (module String) kvs with
+      | `Duplicate_key _ ->
+          Async.Deferred.unit
+      | `Ok _ ->
+          File_system.with_temp_dir "/tmp/coda-test" ~f:(fun db_dir ->
+              let sorted =
+                List.sort kvs ~compare:[%compare: string * string]
+                |> List.map ~f:(fun (k, v) -> (to_bigstring k, to_bigstring v))
+              in
+              let db = create db_dir in
+              List.iter sorted ~f:(fun (key, data) -> set db ~key ~data) ;
+              let alist =
+                List.sort (to_alist db)
+                  ~compare:[%compare: Bigstring.t * Bigstring.t]
+              in
+              [%test_result: (Bigstring.t * Bigstring.t) list] ~expect:sorted
+                alist ;
+              close db ;
+              Async.Deferred.unit ) )
+
+let%test_unit "checkpoint read" =
+  let open Async in
+  Thread_safe.block_on_async_exn
+  @@ fun () ->
+  Quickcheck.async_test ~trials:50
+    Quickcheck.Generator.(
+      list @@ tuple2 String.quickcheck_generator String.quickcheck_generator)
+    ~f:(fun kvs ->
+      match Hashtbl.of_alist (module String) kvs with
+      | `Duplicate_key _ ->
+          Deferred.unit
```

Comment:
maybe dedupe instead of discard? I'm worried about randomly generating trials with dupes and silently not testing this as thoroughly as we hope..

Reply:
Nice catch.